pkg/rancher-desktop/preload/extensions.ts

```diff
@@ -5,21 +5,266 @@
 
 import Electron from 'electron';
 
+import type { SpawnOptions } from '@pkg/main/extensions/types';
+import clone from '@pkg/utils/clone';
 import { ipcRenderer } from '@pkg/utils/ipcRenderer';
 
 /* eslint-disable import/namespace -- that rule doesn't work with TypeScript type-only imports. */
 import type { v1 } from '@docker/extension-api-client-types';
 
+// We use a bunch of symbols for names of properties we do not want to reflect
+// over.
+const stream = Symbol('stream');
+const stdout = Symbol('stdout');
+const stderr = Symbol('stderr');
+const id = Symbol('id');
+
+/** execProcess holds the state associated with a v1.ExecProcess. */
+interface execProcess {
+  /** The identifier for this process. */
+  [id]: string;
+  [stdout]: string;
+  [stderr]: string;
+  [stream]: v1.ExecStreamOptions;
+}
+
+/**
+ * The identifier for the extension (the name of the image).
+ */
+const extensionId = decodeURIComponent((location.href.match(/:\/\/([^/]+)/)?.[1] ?? '').replace(/(..)/g, '%$1'));
+
+/**
+ * The processes that are waiting to complete, keyed by the process ID.
+ * This uses weak references so that if the user no longer cares about them we
+ * will not either.
+ */
+const outstandingProcesses: Record<string, WeakRef<execProcess>> = {};
+
+/**
+ * Construct a TypeError message that is similar to what the browser would
+ * have constructed.
+ * @param name The name of the argument.
+ * @param expectedType: The name of the type that was expected.
+ * @param object The actual object that was passed in (of the incorrect type).
+ */
+function getTypeErrorMessage(name: string, expectedType: string, object: any) {
+  let message = `[ERROR_INVALID_ARG_TYPE]: The "${ name }" argument must be of type ${ expectedType }.`;
+
+  if (typeof object === 'object' && 'constructor' in object && 'name' in object.constructor.name) {
```

Comment:
I think you mean `'name' in object.constructor` or maybe just `object?.constructor.name`
```